environments/servlet/core/src/main/java/org/jboss/weld/environment/undertow/WeldServletExtension.java

```diff
@@ -54,7 +58,10 @@ public void handleDeployment(DeploymentInfo deploymentInfo, ServletContext servl
             // Listener injection
             for (ListenerInfo listener : deploymentInfo.getListeners()) {
                 UndertowLogger.LOG.installingCdiSupport(listener.getListenerClass());
-                listener.setInstanceFactory(WeldInstanceFactory.of(listener.getInstanceFactory(), servletContext, listener.getListenerClass()));
+                InstanceFactory<? extends EventListener> instanceFactory = listener.getInstanceFactory();
+                if (!(instanceFactory instanceof ImmediateInstanceFactory)) {
```

Comment:
@darranl  so, when exactly is `ImmediateInstanceFactory` used? Looking at UT code (which I don't know so I might be wrong), it looks like it's used for any dynamic component. That would mean this change is rather aggressive and will result in leaving all such instance creation to UT instead of trying to let Weld do it. Or am I missing something?

Reply:
Outside of test cases the only class I see actually using this is ServletContextImpl which used it in three locations.

https://github.com/undertow-io/undertow/blob/master/servlet/src/main/java/io/undertow/servlet/spec/ServletContextImpl.java#L540
https://jakarta.ee/specifications/platform/8/apidocs/javax/servlet/servletcontext#addServlet-java.lang.String-javax.servlet.Servlet-

https://github.com/undertow-io/undertow/blob/master/servlet/src/main/java/io/undertow/servlet/spec/ServletContextImpl.java#L632
https://jakarta.ee/specifications/platform/8/apidocs/javax/servlet/servletcontext#addFilter-java.lang.String-javax.servlet.Filter-

https://github.com/undertow-io/undertow/blob/master/servlet/src/main/java/io/undertow/servlet/spec/ServletContextImpl.java#L734
https://jakarta.ee/specifications/platform/8/apidocs/javax/servlet/servletcontext#addListener-T-

So on the Undertow side each of these three points the API is registering instances of the relevant types, as the API talks about instances I don't think it is appropriate to be discarding and recreating the instances.

Within the subsystem this is used within WebHostService where it is checked if an instance already exists to be used:

https://github.com/wildfly/wildfly/blob/master/undertow/src/main/java/org/wildfly/extension/undertow/WebHostService.java#L81

Then the other place it is used is UndertowDeploymentInfoService which is where the error being investigated originated.

Firstly:

https://github.com/wildfly/wildfly/blob/master/undertow/src/main/java/org/wildfly/extension/undertow/WebHostService.java#L81
The LogoutSessionListener here has been passed an instance of AuthentiationManager so would not be correct to recreate.

https://github.com/wildfly/wildfly/blob/master/undertow/src/main/java/org/wildfly/extension/undertow/deployment/UndertowDeploymentInfoService.java#L889
I haven't traced this one but the instances of ServletContainerInitializer already exist so it feels it would be odd to recreate.

https://github.com/wildfly/wildfly/blob/master/undertow/src/main/java/org/wildfly/extension/undertow/deployment/UndertowDeploymentInfoService.java#L1080
This is another instance relying on existing state so recreating if it was possible would loose access to that state.


Overall all the instances I can find the ImmediateInstanceFactory is wrapping a previously instantiated instance which potentially has it's own state and in some cases backed by the API documentation which appears to be talking about registering instances.